### DISCUSSION TITLE
Modules: New module for connecting external ETL Engines

### DISCUSSION THREAD

**Discussion by cong78** [fybrik/fybrik#389] [Ideas]
The Mesh for Data will potentially connect to different ETL engines to perform data actions based on the workload instructions. One of the candidate is IBM Data Stage that is being used in ING as the standard ETL engine of Data Lake. Therefore, a generic ETL Engine Module can be used to create clients for calling Data Stage APIs for performing any ETL actions with results back to M4D. 

The first use case flow would be Data Owner define the data ingestion request to M4D and call Data Stage API for executing predefined job for loading data from a bucket as the System of Records to a target bucket in a Data Lake. Once the job is finished it will notify this ETL Engine client with job status. 

More details will be added soon and any questions are welcomed.

**Comment by roee88** [comment]
Hi @cong78,

Great to see you here. 

Is that a copy module implemented using data stage? What do you mean by generic in "a generic ETL Engine Module"? 

Also, are you aware of the work @eletonia is doing regarding an hello world copy module? https://github.com/IBM/the-mesh-for-data/discussions/255. Just checking if the plan is to leverage her work for the module definition and focus on the actual business logic of the module.

**Reply by cong78** [reply]
hi @roee88 yes the idea is to use the REST APIs from Data Stage that when there is a new request from M4D it can trigger the call of Data Stage with its predefined data stage jobs with relevant inputs. 

For calling a generic ETL engine module because there might be other ETL tooling can perform the similar tasks of what Data Stage can do. So maybe when there is another integration proposal, we can reuse this module to do similar requests, such as run the job, get job status, stop the job, etc. The difference would be during the configuration or client creation phase, which ETL engine will be chosen based on what kind of underlying ETL data objects. 

Yes Sima has told me the hellp world example and I will definitely use that as my hello world starting point. :)

**Reply by ronenkat** [reply]
(moving the discussion to this thread)
@cong78 can you list the additional parameters you need for the ETL module? can you explain how their value would be determined? for example, what is the relationship between the data asset, project and job name? can they be computed using the asset name?
@shlomitk1 what custom properties can we pass through m4dapplication? or through metadata/policies?

**Reply by shlomitk1** [reply]
@ronenkat We can pass any properties through m4dapplication which are dataset independent. Currently, these are role and intent used for evaluating policies. We could add a set of key:value properties defined in m4dapplication to modules (currently, they are not transferred to modules). For example, periodic update could be such property, when all data sets are ingested together with the same periodicity.

**Comment by cong78** [comment]
Hi @shlomitk1 & @ronenkat here is the list of the parameters to reach out to the predefined Data Stage job. 

Parameter | Description
-- | --
service tier host name | The service address where the data stage is hosted. 
jobName | The name of the job that you want to run. It must be a parallel engine or sequence job.jobName=DSJOB1
projectName | The name of the project that contains the job that you want to run.projectName=DSTAGE1
hostName | The host name of the computer where the engine tier is installed.
Datastage Username | User name for authentication. 
Datastage Password | Password for login. 

The example would be here: 
[Documentation](https://www.ibm.com/support/knowledgecenter/en/SSZJPZ_11.7.0/com.ibm.swg.im.iis.ds.fd.doc/topics/rest_api.html#run)

**Reply by cong78** [reply]
@shlomitk1 & @ronenkat Can I reuse the interface part for DataStage parameters?

Original version:
data:
    - dataSetID: "{\"ServerName\":\"cocoMDS3\",\"AssetGuid\":\"ASSET_ID\"}"
      requirements:
        copy:
          required: true
        interface:
          protocol: m4d-arrow-flight
          dataformat: parquet

With Data Stage: 
data:
    - dataSetID: "{\"ServerName\":\"cocoMDS3\",\"AssetGuid\":\"ASSET_ID\"}"
      requirements:
        copy:
          required: true
        interface:
          protocol: data-stage
          hostName: ABC
          jobName: Ingestion
          projectName: Test

**Reply by shlomitk1** [reply]
@cong78 Some clarification questions:
- What is the relation between m4dapplications, datasets, projects, jobs, hosts and services? Assuming I want to create two m4dapplications (app1 and app2), each having two datasets (d11,d12,d21,d22). How many projects/jobs/hosts/services will be created? 
- What is created as a part of the module orchestration and what needs to be pre-installed? For example, is a job created by the module, or does module connect to the existing job?

I'm trying to understand what a module controls (e.g. creates a job) and what information it must be provided during installation? in run-time? (e.g. host, service). If a module is creating a job per dataset, then jobName can be created automatically using release name and hash(dataset). 
If a single host/service  is used, it can be configured inside a module as a part of control-plane installation, like we configure where OPA runs.

**Reply by ronenkat** [reply]
@cong78, following on @shlomitk1 's questions, and looking at the parameters you specified.
Can we assume that service tier and host name are a configuration parameter - they are defined when you install a m4d ETL module? That is, we have a well known Data Stage server running, and we install a module that uses that specific DS instance.

jobname and projectname are already defined when we run the M4DApplication, and is very related to the specific ETL process we wants to run. Do you see these as tied with an asset? i.e., for the Transactions table there is a tag that says which project/job is responsible for ETL? or the tag provide input for a policy that decide what the values should be?

**Comment by simanadler** [comment]
@cong78 @shlomitk1 @ronenkat The intent wasn’t to pass info about Data Stage jobs via M4DApplication but rather to have a parameterized Data Stage job that receives the source and destination from M4D. The name of the job can be defined in the Data Stage module, so that it doesn't have to be hard coded in the code itself.

For multi-cluster, the assumption would then be that the job name would be the same on all the different clusters.  M4D chooses the cluster, and thus the relevant Data Stage instance.

**Comment by cong78** [comment]
@shlomitk1 @ronenkat @simanadler 
That's a very good question. I think I need to make a clarification on what would be the responsibility of ETL Engine Module and what it should derive from M4Dapplication.

In the ideal situation, when there is a data ingestion request, M4D should be able to identify the source and destination country/domain for the ingestion and then ETL Engine Module can consume this information to find the relevant Data Stage instance to perform the ETL jobs in these different countries. The ETL jobs are predefined as templates and can be filled in with actual source and destination information of which dataset it needs to ingest. Therefore I agree the Data Stage host name, Project name, and Job name, etc should be configured during the installation. I will keep them as part of module specific configuration file then. 

But for the current ETL Engine Module I cannot really configure the source and destination parameters for the dataset with existing Data Stage APIs, which means those information has been hard coded in the job already. The only thing I can do is to validate the source and destination from M4Dapplication.yaml to the DataStage compiling logs in XML. Otherwise when there is a request received for copying data, it will just run the job directly.

**Reply by shlomitk1** [reply]
@cong78 Did I understand correctly that there is no way to configure a job that copies 'source' to 'destination' when 'source' and 'destination' are inputs to the module, but rather find an already existing job that is supposed to do so and run it? I've gone over the documentation and what I'm missing there is a CreateJob functionality...

**Reply by cong78** [reply]
> @cong78 Did I understand correctly that there is no way to configure a job that copies 'source' to 'destination' when 'source' and 'destination' are inputs to the module, but rather find an already existing job that is supposed to do so and run it? I've gone over the documentation and what I'm missing there is a CreateJob functionality...

Yes you are right. Currently I cannot find any available methods to do that and I can imagine having such APIs are also difficult in real enterprise environment as they are very complex and not only a one to one copy. 

For current pilot I have to define the job with source and destination parameters in the property field and then run it by the module.

**Reply by simanadler** [reply]
@cong78 @shlomitk1 I spoke to my Data Stage expert and she said that we could define job parameters for sources/target and then provide via the Jobparameter option in the API we are using.   https://www.ibm.com/support/knowledgecenter/en/SSZJPZ_11.7.0/com.ibm.swg.im.iis.ds.fd.doc/topics/rest_api.html#run.  @cong78 This is the field I pointed out in our call on Monday.

**Reply by cong78** [reply]
@simanadler cool I will try it now.

**Reply by simanadler** [reply]
@shlomitk1 @cong78 The destination should be defined by M4D manager - i.e. dynamically allocated.  The source is put in the data owner's private catalog by the data owner, and then he puts the catalog ID in M4DApplication.  M4D then passes that and the destination to the module.

**Reply by shlomitk1** [reply]
@simanadler Right, I was missing the job configuration part (whether M4D controls the target bucket, or it has to be configured in a pre-defined job).

**Reply by cong78** [reply]
@simanadler, @ronenkat  & @shlomitk1 I have a good news and I just tested and it is working by putting those things into Data Stage job parameters as the payload when calling the API, which means we can consume and configure the source and destination info from m4dapplication. The only thing has to be done is all those parameters have to be manually defined and configured when designing the job. I will show more details during the afternoon call.

**Comment by simanadler** [comment]
@cong78 Can you post here the link to the module you created?  I think it's a great example for others.  Thanks!
